### PR TITLE
CDAP-2335 method added to BatchSource/BatchSink that will be called in t...

### PR DESCRIPTION
**Pull request by gokulavasan**
...he onFinish of the ETLMapReduce's onFinish method

JIRA : https://issues.cask.co/browse/CDAP-2335

Bamboo : http://builds.cask.co/browse/CDAP-RT25

To do:
- [x] Add Unit Test

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/templates/etl/api/batch/BatchSink.java [hunk 1]
```diff
@@ -28,6 +28,10 @@
  * also transforms a single input object into a key value pair that the Batch job will output. By default, the input
  * object is used as the key and null is used as the value.
  *
+ * {@link BatchSink#initialize}, {@link BatchSink#transform} and {@link BatchSink#destroy} methods are called inside
+ * the Batch Job while {@link BatchSink#prepareJob} and {@link BatchSink#teardownJob} methods are called on the client
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/templates/etl/api/batch/BatchSink.java]: We don't use `Job`.

Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/templates/etl/api/batch/BatchSink.java]: Removed.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/templates/etl/api/batch/BatchSink.java [hunk 2]
```diff
@@ -49,7 +53,7 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
   public abstract void prepareJob(BatchSinkContext context) throws Exception;
 
   /**
-   * Initialize the sink. This is called once each time the Hadoop Job runs, before any
+   * Initialize the sink. This is called once each time the Batch Job runs, before any
```
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/templates/etl/api/batch/BatchSink.java]: Added BatchEndPointStage.

**Comment by gokulavasan** [comment]
@chtyim Addressed your comments. Please take a look again.

**Comment by gokulavasan** [comment]
Build passed : http://builds.cask.co/browse/CDAP-RT25-11

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/template/etl/api/StageLifecycle.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+import co.cask.cdap.api.annotation.Beta;
+
+/**
+ * Interface for stage that supports initialize call for resource setup.
+ *
+ * @param <T> execution context
+ */
+@Beta
+public interface StageLifecycle<T extends TransformContext> extends Destroyable {
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/template/etl/api/StageLifecycle.java]: The interface name suggest that it's generic, but the context type is not, it is Transform specific.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/template/etl/api/Transformation.java [hunk 4]
```diff
@@ -23,9 +23,10 @@
  *
  * @param <IN> Type of input object
  * @param <OUT> Type of output object
+ * @param <T> execution context
  */
 @Beta
-public interface Transformation<IN, OUT> {
+public interface Transformation<IN, OUT, T extends TransformContext> extends StageLifecycle<T> {
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/template/etl/api/Transformation.java]: Why `Transformation` always need lifecycle? Seems like it should be `Transform` who implements lifecycle.

**Comment by gokulavasan** [comment]
@chtyim Addressed comments. Please take a look. Thanks!

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/template/etl/api/batch/BatchEndPointStage.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import co.cask.cdap.template.etl.api.EndPointStage;
+import co.cask.cdap.template.etl.api.PipelineConfigurer;
+
+/**
+ * EndPointStage for Batch Source/Sink.
+ *
+ * @param <T> batch execution context
+ */
+@Beta
+public abstract class BatchEndPointStage<T extends BatchContext> implements EndPointStage {
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/template/etl/api/batch/BatchEndPointStage.java]: What is EndPoint? Usually we say Http endpoint.

Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/template/etl/api/batch/BatchEndPointStage.java]: It is to refer to the stages at the beginning or the end - ie, source or sink. Any suggestions for better name?

Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/template/etl/api/batch/BatchEndPointStage.java]: I can maybe rename it to EdgeStage

**Comment by gokulavasan** [comment]
@chtyim Renamed EndPointStage -> PipelineConfigurable; BatchEndPointStage -> BatchConfigurable. Please review. Thanks!

**Comment by chtyim** [comment]
LGTM

**Comment by sreevatsanraman** [comment]
Build passed. Merging.

**Pull request merged**